soc: Add support for LPC11U6X soc series

mbittan opened the pull request:
This pull request adds support for NXP LPC11U6X soc series which is embedded in a lot a LaCie/Seagate devices. It includes drivers for clock control, pinmux, gpio, uart, i2c and EEPROM devices and adds support for the NXP LPCXPRESSO11U68 evaluation board based on the LPC11U68JBD100 MCU.

Pinmuxing for I2C and UART devices is handled directly by the corresponding driver instance, which get the pinmux information from the device tree.

This series of patch has been tested on both the LPCXPRESSO11U68 board and another board based on the LPC11U67 MCU.

mbittan committed:
soc: arm: Add basic support for LPC11U6X series

This commit adds basic support for the NXP lpc11u6x soc series.

Signed-off-by: Maxime Bittan <maxime.bittan@seagate.com>

mbittan committed:
boards: arm: add support for lpcxpresso11u68 board

This patch adds support for the LPC11U68 LPCXpresso board based on the
LPC11U68 SoC.

Signed-off-by: Maxime Bittan <maxime.bittan@seagate.com>
Signed-off-by: Simon Guinot <simon.guinot@seagate.com>

mbittan committed:
dts: arm: nxp: Add node and bindings for lpc11u6x clock controller

This commit adds a bindings for the clock controller used in lpc11u6x
MCUs. It also adds a clock-controller node for the lpc11u6x soc series.

Signed-off-by: Maxime Bittan <maxime.bittan@seagate.com>

mbittan committed:
drivers: clock_control: lpc11u6x: Add support for clock controller

This commit adds basic support for the clock controller used in
lpc11u6x MCUs.

Signed-off-by: Maxime Bittan <maxime.bittan@seagate.com>
Signed-off-by: Simon Guinot <simon.guinot@seagate.com>

simonguinot committed:
dts: lpc11u6x: add pinmux node and binding

This patch adds the DT nodes and binding for the pinmux registers,
aka IOCON (I/O control), found on the LPC11U6x MCUs.

Signed-off-by: Simon Guinot <simon.guinot@seagate.com>

simonguinot committed:
drivers: pinmux: sort entries in CMakeLists.txt and Kconfig

Sort source entries in CMakeLists.txt and Kconfig by alphabetic order.

Signed-off-by: Simon Guinot <simon.guinot@seagate.com>

simonguinot committed:
drivers: pinmux: add driver for NXP LPC11U6X MCUs

This patch adds a pinmux driver allowing to configure the IOCON (I/O
control) registers found on the LPC11U6x MCUs.

Signed-off-by: Simon Guinot <simon.guinot@seagate.com>

simonguinot committed:
dts: lpc11u6x: add GPIO node and binding

This patch adds the DT nodes and binding for the GPIO registers found
on the LPC11U6x MCUs.

Signed-off-by: Simon Guinot <simon.guinot@seagate.com>

simonguinot committed:
drivers: gpio: add driver for NXP LPC11U6X MCUs

This patch adds a GPIO driver allowing to configure the GPIOs found on
the LPC11U6x MCUs.

Signed-off-by: Simon Guinot <simon.guinot@seagate.com>

mbittan committed:
dts: arm: Add UART node and binding for LPC11U6X soc series

This commit adds the UART nodes in the LPC11U6X device tree,
as well as its binding.

Signed-off-by: Maxime Bittan <maxime.bittan@seagate.com>

mbittan committed:
drivers: serial: Add support for LPC11U6X UART controllers

This commit adds support for UART0-4 controllers present on
LPC11U6X MCUs.

Signed-off-by: Maxime Bittan <maxime.bittan@seagate.com>

simonguinot committed:
dts: lpc11u6x: add node and binding for on-chip EEPROM

This patch adds the DT nodes and binding for the on-chip EEPROM found
on the LPC11U6x MCUs.

Signed-off-by: Simon Guinot <simon.guinot@seagate.com>

simonguinot committed:
drivers: eeprom: add driver for NXP LPC11U6X

This patch adds adds an EEPROM driver supporting the on-chip EEPROM
found on NXP LPC11U6X MCUs. Note that this driver is only a wrapper
relying entirely on the IAP (In-Application Programming) EEPROM
commands.

Signed-off-by: Simon Guinot <simon.guinot@seagate.com>

simonguinot committed:
soc: nxp_lpc: lpc11u6x: add IAP interface

This patch adds the iap.h header file which provides an entry function
for the IAP (In-Application Programming) interface. Note that the IAP
commands are located in the boot ROM code. Mostly they provide access
to the on-chip flash and EEPROM devices.

Signed-off-by: Simon Guinot <simon.guinot@seagate.com>

mbittan committed:
dts: arm: Add I2C nodes for LPC11U6X soc series

This commit adds the device tree nodes and bindings for the I2C
controllers of the LPC11U6X MCUs

Signed-off-by: Maxime Bittan <maxime.bittan@seagate.com>

mbittan committed:
drivers: i2c: Add support for LPC11U6X I2C controllers

This commit adds a driver for the LPC11U6X I2C controllers

Signed-off-by: Maxime Bittan <maxime.bittan@seagate.com>

galak reviewed (commented):
Are the GPIO, EEPROM, or I2C bits required to boot the board/SoC?  If not I'd ask we split the PR up and have those changes in separate PRs to make it easier to review.

carlescufi commented on soc/arm/nxp_lpc/lpc11u6x/soc.c:
```diff
@@ -0,0 +1,53 @@
+	ARG_UNUSED(arg);
+
+	/* old interrupt lock level */
+	unsigned int oldLevel;
```
Please no camel case

mbittan replied:
Fixed

carlescufi commented on boards/arm/lpcxpresso11u68/doc/index.rst:
```diff
@@ -0,0 +1,164 @@
+
+The LPCXpresso11u68 development board uses an NXP LPC11U68 MCU provides an
+affordable way to prototype new systems.
+
```
Can we remove the marketing speak? Perhaps just mention the MPU

carlescufi commented on boards/arm/lpcxpresso11u68/support/lpc_checksum.py:
```diff
@@ -0,0 +1,123 @@
+#!/usr/bin/env python
+
+# Copyright (c) 2013 Bas Stottelaar
```
@MaureenHelm is it OK to have this Python script here?

mbittan replied:
@henrikbrixandersen @MaureenHelm @carlescufi We added the README file in boards/arm/lpcxpresso11u68/support needed by the new license process.

mbittan replied:
lpc_checksum.py has been added as a dependency in the script/requirements-extra.txt file.

carlescufi commented on drivers/clock_control/clock_control_lpc11u6x.h:
```diff
@@ -0,0 +1,134 @@
+#ifndef ZEPHYR_DRIVERS_CLOCK_CONTROL_CLOCK_CONTROL_LPC11U6X_H_
+#define ZEPHYR_DRIVERS_CLOCK_CONTROL_CLOCK_CONTROL_LPC11U6X_H_
+
+#define LPC11U6X_SYS_AHB_CLK_CTRL_I2C0      (1 << 5)
```
Can’t all those macros come from mcux?

carlescufi commented on drivers/i2c/i2c_lpc11u6x.h:
```diff
@@ -0,0 +1,95 @@
+#ifndef ZEPHYR_DRIVERS_I2C_I2C_LPC11U6X_H_
+#define ZEPHYR_DRIVERS_I2C_I2C_LPC11U6X_H_
+
+#define LPC11U6X_I2C_CONTROL_AA           (1 << 2)
```
Same, can’t we use mcux for this?

carlescufi reviewed (changes requested):
Thanks for the contribution! Not an expert on this platform but have a couple of comments and questions.

MaureenHelm commented on drivers/gpio/Kconfig.lpc11u6x:
```diff
@@ -0,0 +1,31 @@
+# LPC11U6x GPIO configuration options
+
+# Copyright (c) 2020 2020 Seagate Technology LLC
```
Don't need two 2020's. :)

mbittan replied:
Fixed

MaureenHelm commented on dts/arm/nxp/nxp_lpc11u6x.dtsi:
```diff
@@ -6,6 +6,7 @@
 #include <arm/armv6-m.dtsi>
 #include <dt-bindings/clock/lpc11u6x_clock.h>
 #include <dt-bindings/gpio/gpio.h>
+#include <dt-bindings/clock/lpc11u6x_clock.h>
```
Don't need to add this again

mbittan replied:
Fixed

MaureenHelm reviewed (changes requested):
Thank you for the contribution!

Please review the failed checks (codeowners, nits)

mbittan commented:
Thank you everyone for your reviews. I just pushed a new version of the branch that addresses most of the issues that have been raised.
Regarding the license of the script lpc_checksum.py, is there a specific procedure to follow ?

henrikbrixandersen commented on drivers/pinmux/Kconfig:
```diff
@@ -30,28 +30,30 @@ config PINMUX_INIT_PRIORITY
-
 source "drivers/pinmux/Kconfig.beetle"
 
-source "drivers/pinmux/Kconfig.sifive"
-
 source "drivers/pinmux/Kconfig.cc13xx_cc26xx"
 
 source "drivers/pinmux/Kconfig.esp32"
 
-source "drivers/pinmux/Kconfig.sam0"
+source "drivers/pinmux/Kconfig.hsdk"
 
 source "drivers/pinmux/Kconfig.intel_s1000"
 
+source "drivers/pinmux/Kconfig.lpc11u6x"
+
+source "drivers/pinmux/Kconfig.mcux"
+
+source "drivers/pinmux/Kconfig.mcux_lpc"
+
 source "drivers/pinmux/Kconfig.rv32m1"
 
-source "drivers/pinmux/Kconfig.xec"
+source "drivers/pinmux/Kconfig.sifive"
 
-source "drivers/pinmux/Kconfig.hsdk"
```
Could the sorting of this file be submitted as a separate PR?

simonguinot replied:
Do you suggest to make this PR depend on another one sorting the Kconfig / CMakeLists.txt files ? Or do you rather suggest to drop this patch and eventually come back "later" with another PR to do some sorting ?

Honestly, this was the occasion to do some cleaning but we can drop this patch...

pabigot replied:
My personal opinion is that it's OK to mix commits to do cleanup into PRs that subsequently change the file, which is what seems to be happening here.  It's not generally ok to combine cleanup with other changes in the commit , which is not happening here.

But we don't have a policy.

simonguinot replied:
OK let's keep this patch in this PR then.

henrikbrixandersen commented on drivers/serial/CMakeLists.txt:
```diff
@@ -10,6 +10,7 @@ zephyr_library_sources_if_kconfig(uart_esp32.c)
 zephyr_library_sources_if_kconfig(uart_sifive.c)
 zephyr_library_sources_if_kconfig(uart_gecko.c)
 zephyr_library_sources_if_kconfig(leuart_gecko.c)
+zephyr_library_sources_ifdef(CONFIG_UART_LPC11U6X uart_lpc11u6x.c)
```
Why not use `zephyr_library_sources_if_kconfig()`?

mbittan replied:
My bad, we should use it. It is fixed now.

henrikbrixandersen commented on drivers/serial/Kconfig.lpc11u6x:
```diff
@@ -0,0 +1,41 @@
+
+if UART_LPC11U6X
+
+config UART_LPC11U6X_PORT0
+	bool "Enable LPC11U6X UART0"
+	help
+	  Enable UART port 0
+
+config UART_LPC11U6X_PORT1
+	bool "Enable LPC11U6X UART1"
+	help
+	  Enable UART port 1
+
+config UART_LPC11U6X_PORT2
+	bool "Enable LPC11U6X UART2"
+	help
+	  Enable UART port 2
+
+config UART_LPC11U6X_PORT3
+	bool "Enable LPC11U6X UART3"
+	help
+	  Enable UART port 3
+
+config UART_LPC11U6X_PORT4
+	bool "Enable LPC11U6X UART4"
+	help
+	  Enable UART port 4
```
Could these be obtained from devicetree instead of Kconfig?

mbittan replied:
Indeed it can, I removed Kconfig option in favor of device tree macro based on the node's status.

henrikbrixandersen commented on drivers/serial/uart_lpc11u6x.c:
```diff
@@ -0,0 +1,811 @@
+static int lpc11u6x_uart0_configure(struct device *dev,
+				    const struct uart_config *cfg)
+{
+	return -ENOTSUP;
```
Wont this give problems with portable code, which attempts to configure the UART? Would it be better to check the given configuration and only return `-ENOTSUP` if the requested configuration is not applicable for the LPC11U6x?

mbittan replied:
Indeed, I did not think of that. I implemented the configuration functions for values that are supported by LPC11U6X

henrikbrixandersen reviewed (changes requested):
Thank you for this large contribution. It looks really good.

I am a bit divided regarding the split between `soc.c` and the clock controller driver, but since the clock controller driver is very SoC specific, this might be the best solution.

The pinmux design and implementation looks very interesting. I would be very interested in hearing @galak if this is in-line with the generic pinmux design plans for Zephyr v2.4 (and if so, if we should start migrating other SoCs/drivers towards this).

simonguinot commented on drivers/eeprom/eeprom_lpc11u6x.c:
```diff
@@ -0,0 +1,121 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT nxp_lpc11u6x_eeprom
```
Hi Henrik,
I don't know how much the IAP is shared between all the LPCs. Depending on the MCUs some feature may be available or not. The entry codes and the return status may be different as well. I think @MaureenHelm is the one able to answer this.

chrta replied:
I just had a quick look at the IAP of the lpc15xx, see https://www.nxp.com/docs/en/user-guide/UM10736.pdf pg. 572. It seems that the command codes, return values etc are the same. Only the iap entry location seems to be different:
```
#define IAP_LOCATION 0x0300 0205
```

simonguinot replied:
> I just had a quick look at the IAP of the lpc15xx, see https://www.nxp.com/docs/en/user-guide/UM10736.pdf pg. 572. It seems that the command codes, return values etc are the same. Only the iap entry location seems to be different:
> 
> ```
> #define IAP_LOCATION 0x0300 0205
> ```

Hi @chrta, Thanks for looking.

So should we move it under soc/arm/nxp_lpc/common right now ? Or shall we rather wait for someone adding support for another LPC SoC needing IAP to do that ?

MaureenHelm replied:
@agansari can you take a look at this?

bbilas replied:
> > I just had a quick look at the IAP of the lpc15xx, see https://www.nxp.com/docs/en/user-guide/UM10736.pdf pg. 572. It seems that the command codes, return values etc are the same. Only the iap entry location seems to be different:
> > ```
> > #define IAP_LOCATION 0x0300 0205
> > ```
> 
> Hi @chrta, Thanks for looking.
> 
> So should we move it under soc/arm/nxp_lpc/common right now ? Or shall we rather wait for someone adding support for another LPC SoC needing IAP to do that ?

I have some devboards based on LPC1549 so if you need someone to make some tests and etc I am at your disposal.

agansari replied:
@mbittan @MaureenHelm the IAP software has the same command structure across older LPC devices; it's used for flash memories most of the time, but in the case if LPC11U6x it's used for eeprom. I see no flash programming in this pull request.
EEPROM driver should be separate from flash, even if they use the same underlying firmware for Zephyr's driver sake, sdk drivers are separate as well.
As bare metal drivers that don't rely on hal_nxp module they are correct and I would approve, but I have not reviewed in detail the other drivers and configurations.
In case flash will be used, a driver that also fixes #6040 and relies on hal/nxp/mcux/drivers/lpc/fsl_flashiap.h would be preferred.

mbittan replied:
@agansari We don't plan on adding support for the flash controller right now, do you think we should keep the EEPROM driver as it is now ?

agansari replied:
@mbittan yes, there is no other LPC device using eeprom with Zephyr support for now. Also:

> EEPROM driver should be separate from flash, even if they use the same underlying firmware for Zephyr's driver sake, sdk drivers are separate as well.

simonguinot commented on drivers/gpio/Kconfig.lpc11u6x:
```diff
@@ -0,0 +1,31 @@
+
+if GPIO_LPC11U6X
+
+config GPIO_LPC11U6X_PORT0
+	bool "Port 0"
+	help
+	  Enable port 0.
+
+config GPIO_LPC11U6X_PORT1
+	bool "Port 1"
+	help
+	  Enable port 1.
+
+config GPIO_LPC11U6X_PORT2
+	bool "Port 2"
+	help
+	  Enable port 2.
```
Yes it can. In my understanding (coming from the Linux world), DT is about describing the hardware, and not about configuring it. In Linux there is a strong line about that. So if a board does provide two GPIO ports, then they have to be defined and enabled both in the DT. And the application can choose to use them, or not... With @mbittan we realized this was maybe not the way it was working in Zephyr. But it was pretty late in the process and we chose to keep the GPIO driver with the CONFIG_ port options and an abnormal macros usages to ensure a correct match between GPIO instances and ports numbers. 

So you confirm that the correct way to disable a GPIO port is by disabling its DT node from an application overlay ?

pabigot replied:
Yes, that's how this is being done these days.  It's closer to the Linux way than Kconfig.  Some day we should have the ability to enable/disable at runtime, too, for power saving.

mbittan replied:
We removed the CONFIG_ options and are now using the status to instantiate a GPIO device.

simonguinot commented on drivers/gpio/gpio_lpc11u6x.c:
```diff
@@ -0,0 +1,655 @@
+		gpio_regs->clr[port_num] |= BIT(pin);
+	}
+
+	/*
+	 * TODO: maybe configure the STARTERP0 register to allow wake-up from
+	 * deep-sleep or power-down modes.
+	 */
```
I think this should come from a flag when configuring a GPIO interrupt.

henrikbrixandersen commented:
> Regarding the license of the script lpc_checksum.py, is there a specific procedure to follow ?

@mbittan Please see https://docs.zephyrproject.org/latest/contribute/index.html#contribute-non-apache

mbittan commented on soc/arm/nxp_lpc/Kconfig:
```diff
@@ -1,4 +1,4 @@
-# Copyright (c) 2017, NXP
+# Copyright (c) 2020, Seagate
```
Fixed

mbittan commented on soc/arm/nxp_lpc/lpc11u6x/Kconfig.soc:
```diff
@@ -0,0 +1,52 @@
+if SOC_SERIES_LPC11U6X
+
+config SOC_PART_NUMBER_LPC11U66JBD48
+       bool
```
Fixed

mbittan commented on boards/arm/lpcxpresso11u68/Kconfig.defconfig:
```diff
@@ -0,0 +1,11 @@
+if BOARD_LPCXPRESSO11U68
+
+config BOARD
+	default "lpcxpresso11u68" if BOARD_LPCXPRESSO11U68
```
Fixed

mbittan commented on soc/arm/nxp_lpc/lpc11u6x/Kconfig.defconfig.series:
```diff
@@ -14,5 +14,11 @@ config NUM_IRQS
+if CLOCK_CONTROL
+
+config CLOCK_CONTROL_LPC11U6X
+       default y
```
Fixed

mbittan commented on boards/arm/lpcxpresso11u68/lpcxpresso11u68.yaml:
```diff
@@ -0,0 +1,17 @@
+testing:
+  ignore_tags:
+    - net
+    - bluetooth
```
Indeed, they are not and have thus been removed.

mbittan commented on boards/arm/lpcxpresso11u68/lpcxpresso11u68_defconfig:
```diff
@@ -0,0 +1,21 @@
+CONFIG_GPIO=y
+CONFIG_CORTEX_M_SYSTICK=y
+CONFIG_SOC_SERIES_LPC11U6X=y
+CONFIG_MAIN_STACK_SIZE=512
+CONFIG_ISR_STACK_SIZE=768
```
Since the board only has 32k of RAM, we decided to lower some of the stack sizes. I added a comment to explain that.

mbittan commented on boards/arm/lpcxpresso11u68/lpcxpresso11u68_defconfig:
```diff
@@ -0,0 +1,21 @@
+CONFIG_SOC_SERIES_LPC11U6X=y
+CONFIG_MAIN_STACK_SIZE=512
+CONFIG_ISR_STACK_SIZE=768
+CONFIG_POLL=y
```
My bad, it shoudn't. It is now fixed.

henrikbrixandersen commented on boards/arm/lpcxpresso11u68/support/lpc_checksum_README:
```diff
@@ -0,0 +1,40 @@
+lpc_checksum
+############
+
+Origin:
+   https://github.com/basilfx/lpc_checksum
+
```
Instead of including the script within the zephyr repo, could we just add it as an (optional) dependency to be installed using `pip install lpc_checksum`?

simonguinot replied:
Yes we can :) When I looked in scripts/requirements*, I missed the "optional" part. Is that possible to install the script only when building Zephyr for a LPC11U6x MCU ? It was not feeling right to me to install this script in everybody's $HOME. That's the reason we chose the direct import. But if a dependency is the right way to go, please confirm and I'll do the change.

simonguinot replied:
What do you think @henrikbrixandersen ?

henrikbrixandersen replied:
I think we should add the dependency to either `scripts/requirements-base.txt` or `scripts/requirements-extras.txt`.

mbittan replied:
The dependency has been added to scripts/requirements-extra.txt

henrikbrixandersen commented on drivers/gpio/gpio_lpc11u6x.c:
```diff
@@ -0,0 +1,655 @@
+		    POST_KERNEL, CONFIG_KERNEL_INIT_PRIORITY_DEVICE,	\
+		    &gpio_lpc11u6x_driver_api)
+
+#if DT_NODE_HAS_STATUS(DT_NODELABEL(gpio0), okay)
+GPIO_LPC11U6X_INIT(0);
+#endif
+
+#if DT_NODE_HAS_STATUS(DT_NODELABEL(gpio1), okay)
+GPIO_LPC11U6X_INIT(1);
+#endif
+
+#if DT_NODE_HAS_STATUS(DT_NODELABEL(gpio2), okay)
+GPIO_LPC11U6X_INIT(2);
+#endif
```
```suggestion
DT_INST_FOREACH_STATUS_OKAY(GPIO_LPC11U6X_INIT)
```

simonguinot replied:
I spent some time this morning trying to do that. The result is I'd rather prefer not :) The way the hardware works, any interruption can be mapped on any GPIO port. So when an interrupt happens, the handler has to retrieve first the GPIO port number and next the "struct device" associated with. To do that, having the GPIO device X associated with the GPIO port X helps a lot. With DT_INST_FOREACH_STATUS_OKAY we can't assume that. If for example the GPIO port 0 node is disabled in DT (let's say by an application overlay), then instance 0 will be GPIO port 1.

henrikbrixandersen commented on drivers/serial/uart_lpc11u6x.c:
```diff
@@ -0,0 +1,996 @@
+		    PRE_KERNEL_1, CONFIG_KERNEL_INIT_PRIORITY_OBJECTS,	      \
+		    &uartx_api)
+
+#if DT_NODE_HAS_STATUS(DT_NODELABEL(uart1), okay)
+LPC11U6X_UARTX_INIT(1);
+#endif /* DT_NODE_HAS_STATUS(DT_NODELABEL(uart1), okay) */
+
+#if DT_NODE_HAS_STATUS(DT_NODELABEL(uart2), okay)
+LPC11U6X_UARTX_INIT(2);
+#endif /* DT_NODE_HAS_STATUS(DT_NODELABEL(uart2), okay) */
+
+#if DT_NODE_HAS_STATUS(DT_NODELABEL(uart3), okay)
+LPC11U6X_UARTX_INIT(3);
+#endif /* DT_NODE_HAS_STATUS(DT_NODELABEL(uart3), okay) */
+
+#if DT_NODE_HAS_STATUS(DT_NODELABEL(uart4), okay)
+LPC11U6X_UARTX_INIT(4);
+#endif /* DT_NODE_HAS_STATUS(DT_NODELABEL(uart4), okay) */
```
```suggestion
DT_INST_FOREACH_STATUS_OKAY(LPC11U6X_UARTX_INIT)
```

mbittan replied:
I'm not sure that would work in my case. On the LPC11U6X MCUs, UART0 has a different interface from the other uart devices, using DT_INST_FOREACH_STATUS_OKAY on LPC11U6X_UARTX_INIT would imply, among other things, initializing all uart devices with the same struct uart_driver_api pointer.
Furthermore, if I understand correctly, there is no guarantee that the instance number Y of a compatible will correspond to the uart Y hardware of the LPC11U6X. In this driver, I need to be sure that the uart device number Y correspond to the actual uart Y of the MCU in order to initialize the shared irq correctly.

mbittan replied:
@henrikbrixandersen Is there a deterministic way to guarantee that instance number X will correspond to the real uartX of the MCU or should we just keep it like this ?

henrikbrixandersen replied:
No, there are no guarantees in that regard. Perhaps @galak has input on how to go about this?

mbittan replied:
@galak Do you have any insight on this ?

simonguinot committed:
scripts/requirements: add lpc_checksum script

On some NXP LPC MCUs (such as LPC11U6x), a valid firmware image must
include a checksum in the 7th entry of the ARM Cortex-M0 exception
vector table. This checksum is verified by the bootloader.

The lpc_checksum.py python script allows to insert this checksum into
the firmware images (BIN and HEX formats).

Project page: https://github.com/basilfx/lpc_checksum

Signed-off-by: Simon Guinot <simon.guinot@seagate.com>

MaureenHelm commented on drivers/clock_control/clock_control_lpc11u6x.c:
```diff
@@ -0,0 +1,400 @@
+}
+
+static void syscon_set_pll_src(struct lpc11u6x_syscon_regs *syscon,
+			       u32_t src)
```
The zephyr-specific integer types were recently deprecated in #25679. Please rebase and update to use C99 integer types

agansari reviewed (commented):
need more time to review this pull

agansari commented on soc/arm/nxp_lpc/lpc11u6x/iap.h:
```diff
@@ -0,0 +1,79 @@
+/**
+ * @brief Entry function for IAP commands.
+ */
+static inline int iap_cmd(unsigned int cmd[5])
```
Would it be possible to use already provided functions?
Have a look at modules/hal/nxp/mcux/drivers/lpc/fsl_flashiap.h

agansari replied:
I see your point

agansari commented on drivers/gpio/gpio_lpc11u6x.c:
```diff
@@ -0,0 +1,655 @@
+	 */
+	intpin = pin;
+	if (config->port_num == 1) {
+		intpin += 24;
```
This looks similar to https://github.com/zephyrproject-rtos/zephyr/blob/4b194eb4fc40723bba46acf0fae36d5d1d1e59a8/drivers/gpio/gpio_mcux_lpc.c#L199 

Could this be reused?

simonguinot replied:
Hi Andrei,

Many thanks for spending some time on this PR.

In the gpio_mcux_lpc driver each free interrupt can be attached _one_ time to a pin. It is never detached. In all the dtsi files using the "nxp,lpc-gpio" binding, two GPIOs ports are defined and 8 interrupts lines are split equally between this two port. Basically each GPIO port gets four interrupts attached statically at built time. Once the 4 interrupts lines are consumed/used, it is done.

On the LPC11U6x SoC there are 8 interrupts lines available as well, but 3 GPIOs ports. This doesn't divide equally. In addition each of the 8 interrupts can be used by each of the GPIO ports. So we choose a dynamic approach. At run time any of the port can attach any of the free interrupts lines (as longer there is one available). An when an interrupt is disabled then it is detached from the pin. So the interrupt line becomes available again for any of the GPIO ports. It is done by the pintsel_detach function. In the gpio_mcux_lpc driver and in the NXP MCUx HAL, there is no support to detach interrupts from pin. To support the attach/detach feature the pintsel_attach function have to run some checks on the pintsel registers. attach_pin_to_isr() does not. That's why I think that attach_pin_to_isr should not be reused here.

agansari replied:
Thank you Simon for your answer. You are right, pins are not changed at runtime, once attached they are not detached, this comes from allocating resources statically in device tree. If you need to detach the pin at runtime it wouldn't be a good ideea to switch to the other driver.
The allocation of 4 interrupt lines/port is arbitrary and should not be taken as hardwired, most of the code in gpio_mcux_lpc is designed so interrupts can be differently distributed depending on application. At the time I implemented that I had LPC55s69 an LPC54114 in mind, the allocation is fair and with no particular application in mind.

agansari reviewed (changes requested):
I have some questions and remarks.
Looks fine as standalone baremetal drivers, but I think some existing code could be reused from other LPC devices.

MaureenHelm reviewed (approved):
Will you please investigate the buildkite failure?

simonguinot commented:
> Will you please investigate the buildkite failure?

Hi Maureen,

Thanks for the approval !

I am not sure how we can fix this failure. Since we have moved the lpc_checksum script into scripts/requirements-extras.txt, then it must be installed in the containers used by the CI.

MaureenHelm commented:
> I am not sure how we can fix this failure. Since we have moved the lpc_checksum script into scripts/requirements-extras.txt, then it must be installed in the containers used by the CI.

@galak can you help with this?

henrikbrixandersen commented:
> I am not sure how we can fix this failure. Since we have moved the lpc_checksum script into scripts/requirements-extras.txt, then it must be installed in the containers used by the CI.

@simonguinot You could add it to `install-packages` in `.github/workflows/sanitycheck_tests.yml`.

mbittan committed:
.buildkite: run.sh: Install lpc_checksum package

The sanitycheck tests are all failing for the lpcxpresso11u68 board
because the lpc_checksum, a new python script requirement introduced
by this board, is not present in the build environment. This
commit attempt to fix this issue by installing lpc_checksum before
running sanitycheck.

Signed-off-by: Maxime Bittan <maxime.bittan@seagate.com>

mbittan commented:
> > I am not sure how we can fix this failure. Since we have moved the lpc_checksum script into scripts/requirements-extras.txt, then it must be installed in the containers used by the CI.
> 
> @galak can you help with this?

I added a pip3 install command in the .buildkite/run.sh script, as suggested by @galak on slack. It fixes the build failure, but it is only a temporary fix.
